draft-irtf-cfrg-bls-signature.md

```diff
@@ -0,0 +1,1290 @@
+
+Outputs:
+- result, either VALID or INVALID.
+
+Procedure:
+1. If any two input messages are equal, return INVALID.
+2. return CoreAggregateVerify((PK_1, message_1), ..., (PK_n, message_n),
+                              signature)
+~~~
+
+## Message augmentation {#schemeaug}
+
+In a message augmentation scheme, signatures are generated
+over the concatenation of the public key and the message,
+ensuring that messages signed by different public keys are
+distinct.
+
+All public keys PK supplied as arguments to Verify and AggregateVerify
+MUST satisfy KeyValidate(PK) == VALID ((#keyvalidate)).
+
+### Sign
+
+To match the API for Sign defined in (#blsapi), this function
+recomputes the public key corresponding to the input SK.
+Implementations MAY instead implement an interface that takes
+the public key as an input.
+
+Note that the point P and the point\_to\_pubkey function are
+defined in (#coreparams).
+
+~~~
+signature = Sign(SK, message)
+
+Inputs:
+- SK, a secret key output by an invocation of KeyGen.
+- message, an octet string.
+
+Outputs:
+- signature, an octet string.
+
+Procedure:
+1. xP = SK * P
+2. PK = point_to_pubkey(xP)
+3. return CoreSign(SK, PK || message)
+~~~
+
+### Verify
+
+~~~
+result = Verify(PK, message, signature)
+
+Inputs:
+- PK, a public key in the format output by KeyGen.
+- message, an octet string.
+- signature, an octet string in the format output by CoreSign.
+
+Outputs:
+- result, either VALID or INVALID.
+
+Procedure:
+1. return CoreVerify(PK, PK || message, signature)
+~~~
+
+### AggregateVerify
+
+~~~
+result = AggregateVerify((PK_1, message_1), ..., (PK_n, message_n),
+                         signature)
+
+Inputs:
+- PK_1, ..., PK_n, public keys in the format output by KeyGen.
+- message_1, ..., message_n, octet strings.
+- signature, an octet string output by Aggregate.
+
+Outputs:
+- result, either VALID or INVALID.
+
+Procedure:
+1. for i in 1, ..., n:
+2.    mprime_i = PK_i || message_i
+3. return CoreAggregateVerify((PK_1, mprime_1), ..., (PK_n, mprime_n),
+                              signature)
+~~~
+
+## Proof of possession {#schemepop}
+
+
+A proof of possession scheme uses a separate public key
+validation step, called a proof of possession, to defend against
+rogue key attacks.
+This enables an optimization to aggregate signature verification
+for the case that all signatures are on the same message.
+
+The Sign, Verify, and AggregateVerify functions
+are identical to CoreSign, CoreVerify, and CoreAggregateVerify
+((#coreops)), respectively.
+In addition, a proof of possession scheme defines three functions beyond
+the standard API ((#blsapi)):
+
+- PopProve(SK) -> proof: an algorithm that generates a proof of possession
+  for the public key corresponding to secret key SK.
+
+- PopVerify(PK, proof) -> VALID or INVALID:
+  an algorithm that outputs VALID if proof is valid for PK, and INVALID otherwise.
+
+- FastAggregateVerify(PK\_1, ..., PK\_n, message, signature) -> VALID or INVALID:
```

Comment:
I'd prefer to go with the point of view that says all the specifications refer to input-output behavior, written in an algorithmic language, and that implementations can adopt algorithmic short-cuts.

From that point of view, FastAggregateVerify would be an algorithmic short-cut, so we don't need an explicit function. (That said, I also see an argument for an explicit function for oft-used algorithmic short-cuts.)

We can have a separate section describing various algorithmic short-cuts. In the case of verification, the short-cuts include:

* using a pairing-product check instead of computing many individual pairings.
* in CoreAggregateVerify, if the messages are the same, then we can multiply the signatures. more generally, we can batch the messages into groups of identical messages, etc...

Reply:
Sure, that's fair. The prior version of the document had this, so I kept it for the PoP case (since that's the only time it applies).

A couple things to note:

1. Pairing-product check requires breaking the `e` abstraction. Probably there's a graceful way to do this (in a non-normative section, for example); my point is just that it's a qualitatively different kind of optimization.

2. The optimization you suggest for CoreAggregateVerify is explicitly insecure except in the PoP case, which is why I didn't consider adding it. That, as you know, is why we have the message uniqueness check in the "basic" AggregateVerify routine. Meanwhile, in message-augmentation, we're essentially guaranteed that there will never be two colliding messages (under the quite reasonable assumption that pubkeys are unique).

    So it seems like the only place this optimization applies is in PoP, just as we've got now. We can certainly move it elsewhere, but I don't think it should go into CoreAggregateVerify.

One other reason to mention the possibility of this optimization in the PoP subsection is that it makes (somewhat) clearer why one would choose PoP over Aug. But maybe we should make this explicit, e.g., with a section that gives some guidance about which scheme to use for what application.